content/general/concepts/xml/terms/xpath/xpath.md

```diff
@@ -0,0 +1,116 @@
+
+Location paths can be used inside predicates, and are evaluated relative to the node being tested. so the `./@C` in `A[./@C="1"]` is looking to match the `C` attribute of the `A` element.
+
+The following operators can be used inside of predicates in XPath 1.0:
+
+- `|`, the union operator creates the union of two node sets.
+- the boolean operators `and` & `or`.
+- the boolean function `not()`
+- mathematical operators: `+`,`-`,`*`,`div` (divide) and `mod`.
+- the comparison operators `=`,`!=`,`<`,`>`,`<=` & `>=`
+
+Predicates also have a number of functions available in XPath 1.0, the following are some of the more commonly used ones:
+
+- `contains(string1, string2)`: Checks if `string2` is contained within `string1` and returns true or false.
+- `count(node-set)`: Returns the number of nodes in `node-set`.
+- `false()`: Returns false.
+- `last()`: Returns the position of the last match of the current node test.
+- `local-name(node-set)`: Returns the local name of the first node in `node-set`, or if `node-set` is omitted, the local name of the current node.
+- `normalize-space(string)`: Returns `string` with leading and trailing whitespace removed, and all sequences of whitespace characters replaced by a single space. If omitted, operates on the context node converted to a string.
+- `position()`: Returns the numeric position of the current node.
+- `starts-with(string1, string2)`: Checks if `string1` starts with `string2` and returns true or false.
+- `sum(node-set)`: Converts the string values of the nodes in `node-set` into numbers and returns the sum of those numbers.
+- `true()`: Returns true.
+
+## Examples
+
+Consider the following XML document:
+
+```pseudo
+<A>
+  <B C="1">
+    <D E="2"/>
+  </B>
+  <B C="3">
+    <D F="4"/>
+  </B>
+</A>
+```
+
+Example XPath expressions include:
+
+- `A/B/D`: The most basic type of XPath expression. Each location step is separated by a slash. The default axis is the child axis and `A`,`B` & `C` represent the element nodes we're matching. This expression matches all `D` elements that are children of `B` elements that are children of `A` elements. It would match both `D` elements in the example. The expression can be written in full expanded form as `child::A/child::B/child::D`
+- `A//D`: Also matches both `D` elements in the example, because the double slash (`//`) represents the descendant-or-self axis. This expression matches all `D` elements that are descendants of the `A` element. The expanded form of the expression is `child::A/descendant-or-self::node()/child::D`
+- `A//D[1]`: Adds a predicate (in the square brackets `[...]`) to the `D` node test. Using a number alone in the square brackets is a shorthand way of specifying the position of the element. This predicate specifies that we're only matching the first `D` descendant of `A`. The expanded form of the expression is `child::A/descendant-or-self::node()/child::D[position()=1]`
```

Comment:
```suggestion
- `A/B/D`: The most basic type of XPath expression. Each location step is separated by a slash. The default axis is the child axis and `A`,`B` & `C` represent the element nodes we're matching. This expression matches all `D` elements that are children of `B` elements that are children of `A` elements. It would match both `D` elements in the example. The expression can be written in full expanded form as `child::A/child::B/child::D`.
- `A//D`: Also matches both `D` elements in the example, because the double slash (`//`) represents the descendant-or-self axis. This expression matches all `D` elements that are descendants of the `A` element. The expanded form of the expression is `child::A/descendant-or-self::node()/child::D`.
- `A//D[1]`: Adds a predicate (in the square brackets `[...]`) to the `D` node test. Using a number alone in the square brackets is a shorthand way of specifying the position of the element. This predicate specifies that we're only matching the first `D` descendant of `A`. The expanded form of the expression is `child::A/descendant-or-self::node()/child::D[position()=1]`.
```